### PR TITLE
Stop storing type information in metadata cache for numeric metrics

### PR DESCRIPTION
**Pull request by tilogaat**
This is one of the children of PR #412 and is the successor of PR #419 and #421 

The idea is to not store the Data Types (I for Int, L for Long etc) The data types are used only for determining if serializers for String and Boolean metrics. For numeric metrics, the serializers are self describing. Wherever, we discover DataType is null, we assume the DataType Numeric.

### CODE REVIEW

File: blueflood-http/src/main/java/com/rackspacecloud/blueflood/inputs/handlers/HttpMetricsIngestionServer.java [hunk 1]
```diff
@@ -172,7 +172,6 @@ private void buildProcessingChains() {
 
         this.defaultProcessorChain = AsyncChain
                 .withFunction(typeAndUnitProcessor)
-                .withFunction(rollupTypeCacher)
```
Comment (dlobue) [inline blueflood-http/src/main/java/com/rackspacecloud/blueflood/inputs/handlers/HttpMetricsIngestionServer.java]: w00t!

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/AstyanaxReader.java [hunk 2]
```diff
@@ -473,7 +485,15 @@ private MetricData transformColumnsToMetricData(Locator locator, ColumnList<Long
...
-            DataType dataType = new DataType(metaCache.get(locator, dataTypeCacheKey));
+            DataType dataType = getDataType(locator, dataTypeCacheKey);
+            if (rollupType == null) {
+                rollupType = RollupType.BF_BASIC;
+            }
+
+            if (dataType == null) {
+                dataType = DataType.NUMERIC;
+            }
+
```
Comment (dlobue) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/AstyanaxReader.java]: I've seen this code several times already. Perhaps bake this logic into the `getDataType` and `RollupType.fromString` methods?

Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/AstyanaxReader.java]: Fixed in  348c662
Turns out RollupType.fromString already had that support. Baked it in getDataType
However, in other places they are passed as params into public and protected methods. We need to be sure there that no one is passing nulls.

File: blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/io/AstyanaxReaderIntegrationTest.java [hunk 3]
```diff
@@ -94,4 +97,67 @@ public void testBatchedReads() throws Exception {
...
+    }
+
+    @Test
+    public void testNullRollupType_DoesNotReturn_StringOrBooleanSerializers() {
+        AstyanaxReader reader = AstyanaxReader.getInstance();
+
+        AbstractSerializer serializer = reader.serializerFor(null, DataType.INT, Granularity.MIN_5);
+
+        Assert.assertTrue(serializer != null);
+        Assert.assertFalse(serializer instanceof StringSerializer);
```
Comment (chinmay-gupte) [inline blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/io/AstyanaxReaderIntegrationTest.java]: Is just testing for `serializer instanceOf NumericSerializer` sufficient here?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/AstyanaxReader.java [hunk 4]
```diff
@@ -326,9 +326,14 @@ public static String getType(Locator locator) {
 
     public MetricData getDatapointsForRange(Locator locator, Range range, Granularity gran) {
         try {
+            //TODO: If we stop processing string metrics, we can get rid of this and always return numeric
+            //Questions: Do we care about pre-agg types
```
Comment (chinmay-gupte) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/AstyanaxReader.java]: Has this question been answered? IMO we will be storing rollup types for pre-agg rollup types, so rollup type will never be `null`?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/outputs/formats/MetricData.java [hunk 5]
```diff
@@ -61,6 +61,15 @@ public String toString() {
         }
 
         public static Type from(RollupType rollupType, DataType dataType) {
+            // We no longer store datatype metadata for Numeric datatypes
+            if (dataType == null) {
+                return NUMBER;
+            }
+
+            if(rollupType == null) {
```
Comment (chinmay-gupte) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/outputs/formats/MetricData.java]: NIT: space between if and `(`

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java [hunk 6]
```diff
@@ -98,6 +98,11 @@ public void run() {
             Rollup rollup = null;
             RollupType rollupType = RollupType.fromString((String) rollupTypeCache.get(
                     singleRollupReadContext.getLocator(), MetricMetadata.ROLLUP_TYPE.name().toLowerCase()));
+
+            if (rollupType == null) {
```
Comment (chinmay-gupte) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java]: Will `rollupType` ever be `null` here? Because rollupTypeCache, will return `BF_BASIC` if its not `null` right?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/DataType.java [hunk 7]
```diff
@@ -32,20 +32,21 @@ public DataType(String type) {
...
     public final static DataType BOOLEAN = new DataType("B");
     public final static DataType BIGINT = new DataType("BI");
+    public final static DataType NUMERIC = new DataType("N");
 
     public static DataType getMetricType(Object metricValue) {
         if (metricValue instanceof String) {
             return STRING;
         } else if (metricValue instanceof Integer) {
-            return INT;
+            return NUMERIC;
```
Comment (chinmay-gupte) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/DataType.java]: Can we consolidate these blocks into a single `metricValue instanceOf Number`? Ref: https://github.com/rackerlabs/blueflood/pull/421#discussion_r23633118

File: blueflood-http/src/test/java/com/rackspacecloud/blueflood/inputs/formats/JSONMetricsContainerTest.java [hunk 8]
```diff
@@ -53,7 +53,7 @@ public void testJSONMetricsContainerConstruction() throws Exception {
         Assert.assertEquals(1234566, metricsCollection.get(0).getTtlInSeconds());
         Assert.assertEquals(1234567890L, metricsCollection.get(0).getCollectionTime());
         Assert.assertEquals("milliseconds", metricsCollection.get(0).getUnit());
-        Assert.assertEquals("L", metricsCollection.get(0).getDataType().toString());
+        Assert.assertEquals("N", metricsCollection.get(0).getDataType().toString());
```
Comment (chinmay-gupte) [inline blueflood-http/src/test/java/com/rackspacecloud/blueflood/inputs/formats/JSONMetricsContainerTest.java]: No one uses this data type marker, I think. And the type of the data can be easily referred from the JSON types. But its obviously beyond of the scope of this PR, but we should do it eventually.

File: blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/io/AstyanaxReaderIntegrationTest.java (hunk 3 above)
Comment (tilogaat) [inline blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/io/AstyanaxReaderIntegrationTest.java]: @chinmay-gupte Unfortunately no. serializerFor is a static method in NumericSerializer class that returns something like AbstractSerializer<T>.

Comment (tilogaat) [inline blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/io/AstyanaxReaderIntegrationTest.java]: https://github.com/rackerlabs/blueflood/blob/master/blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/serializers/NumericSerializer.java#L80

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java (hunk 6 above)
Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java]: Fixed in 7c6bd2d

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/outputs/formats/MetricData.java (hunk 5 above)
Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/outputs/formats/MetricData.java]: Fixed 7c6bd2d

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/AstyanaxReader.java (hunk 4 above)
Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/AstyanaxReader.java]: Fixed in 7c6bd2d

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/DataType.java (hunk 7 above)
Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/DataType.java]: Fixed in 2053562

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java [hunk 9]
```diff
@@ -95,11 +93,16 @@ private IncomingMetricException checkMeta(Locator locator, String key, String in
...
 
-        IncomingMetricException typeProblem = checkMeta(metric.getLocator(), MetricMetadata.TYPE.name().toLowerCase(),
-                metric.getDataType().toString());
-        IncomingMetricException unitProblem = checkMeta(metric.getLocator(), MetricMetadata.UNIT.name().toLowerCase(),
-                metric.getUnit());
+        if (metric.getDataType() != DataType.NUMERIC) {
+            typeProblem = checkMeta(metric.getLocator(), MetricMetadata.TYPE.name().toLowerCase(),
+                    metric.getDataType().toString());
+
+            unitProblem = checkMeta(metric.getLocator(), MetricMetadata.UNIT.name().toLowerCase(),
```
Comment (chinmay-gupte) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java]: `checkMeta` for unit does not have to be in `metric.getDataType() != DataType.NUMERIC`, I think. It will be handled as a separate block of code in https://github.com/rackerlabs/blueflood/pull/432

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        ArrayList<IMetric> metricsList = new ArrayList<IMetric>();
+
+        Locator locator = Locator.createLocatorFromDbKey("integer_metric");
+        metricsList.add(new Metric(locator, 123, 14200000, new TimeValue(1, TimeUnit.DAYS),"somethings"));
+
+        analyzer.scanMetrics(metricsList);
+        String type = cache.get(locator, MetricMetadata.TYPE.name().toLowerCase());
+        Assert.assertEquals(null, type);
+
+        String unit = cache.get(locator, MetricMetadata.UNIT.name().toLowerCase());
```
Comment (chinmay-gupte) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java]: This test will be changed, once you remove the `checkMeta` for unit in `IncomingMetadataAnalyzer` code.

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        ArrayList<IMetric> metricsList = new ArrayList<IMetric>();
+
+        Locator locator = Locator.createLocatorFromDbKey("preag_counter");
+        metricsList.add(new PreaggregatedMetric(12300000, locator, new TimeValue(1, TimeUnit.DAYS), new CounterRollup()));
+
+        analyzer.scanMetrics(metricsList);
+        String type = cache.get(locator, MetricMetadata.TYPE.name().toLowerCase());
+        Assert.assertEquals(null, type);
+
+        String unit = cache.get(locator, MetricMetadata.UNIT.name().toLowerCase());
```
Comment (chinmay-gupte) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java]: Ditto, this test will get changed, once you make the suggested change in `IncomingMetadataAnalyzer`.

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        cache.setIO(this.metadataIO);
+        IncomingMetricMetadataAnalyzer analyzer = new IncomingMetricMetadataAnalyzer(cache);
+
+        ArrayList<IMetric> metricsList = new ArrayList<IMetric>();
+
+        Locator locator = Locator.createLocatorFromDbKey("preag_counter");
+        metricsList.add(new PreaggregatedMetric(12300000, locator, new TimeValue(1, TimeUnit.DAYS), new CounterRollup()));
+
+        analyzer.scanMetrics(metricsList);
+        String type = cache.get(locator, MetricMetadata.TYPE.name().toLowerCase());
```
Comment (chinmay-gupte) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java]: Just curious, what type info gets stored for preaggregated metrics before we make this change?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java [hunk 13]
```diff
@@ -52,7 +49,8 @@ public IncomingMetricMetadataAnalyzer(MetadataCache cache) {
         for (IMetric metric : metrics) {
             try {
                 if (metric instanceof Metric) {
-                    Collection<IncomingMetricException> metricProblems = checkMetric((Metric)metric);
+                    DataType datatype = ((Metric) metric).getDataType();
```
Comment (chinmay-gupte) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java]: Unused?

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java [hunk 14]
```diff
@@ -98,6 +98,7 @@ public void run() {
             Rollup rollup = null;
             RollupType rollupType = RollupType.fromString((String) rollupTypeCache.get(
                     singleRollupReadContext.getLocator(), MetricMetadata.ROLLUP_TYPE.name().toLowerCase()));
+
```
Comment (chinmay-gupte) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java]: NIT: extra blank line

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java (hunk 12 above)
Comment (tilogaat) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java]: @chinmay-gupte nothing gets stored. If you see in scanMetrics only instances of Metric are checked. PreaggregatedMetric implements IMetric and not Metric. So before making this change itself, no type info was being stored for PreaggregatedMetrics

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java (hunk 10 above)
Comment (tilogaat) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/inputs/processors/IncomingMetricMetadataAnalyzerTest.java]: Fixed in  7babce5

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java (hunk 9 above)
Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java]: Fixed in  7babce5

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java (hunk 13 above)
Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java]: Fixed in  7babce5

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java (hunk 14 above)
Comment (tilogaat) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java]: Fixed in  7babce5

File: blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java [hunk 15]
```diff
@@ -95,10 +92,15 @@ private IncomingMetricException checkMeta(Locator locator, String key, String in
...
 
-        IncomingMetricException typeProblem = checkMeta(metric.getLocator(), MetricMetadata.TYPE.name().toLowerCase(),
-                metric.getDataType().toString());
-        IncomingMetricException unitProblem = checkMeta(metric.getLocator(), MetricMetadata.UNIT.name().toLowerCase(),
+        if (metric.getDataType() != DataType.NUMERIC) {
+            typeProblem = checkMeta(metric.getLocator(), MetricMetadata.TYPE.name().toLowerCase(),
+                    metric.getDataType().toString());
+        }
+
+        unitProblem = checkMeta(metric.getLocator(), MetricMetadata.UNIT.name().toLowerCase(),
```
Comment (chinmay-gupte) [inline blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/IncomingMetricMetadataAnalyzer.java]: NIT: Revert this change to have `IncomingMetricException unitProblem = checkMeta(...`

**Comment by chinmay-gupte** [comment]
LGTM. +1 after https://github.com/rackerlabs/blueflood/pull/428/files#r24265425 is fixed. Thanks for your patience!

**Comment by tilogaat** [comment]
@chinmay-gupte tearDown method makes sure that we set the private fields back to their original values. 
I tried calling init, but to no use. I think in the first test the static fields are set. I think a better approach would be to get rid of the private fields and just invoke the Configs directly. What do you think?

**Comment by chinmay-gupte** [comment]
+1
